File: examples/rust/src/notification.rs
// Copyright 2026 Chengxi Luo
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! `NotificationRecord` mirrors examples/cases/notification.yaml, whose
//! `channel` field is a `oneof`.

use serify::SerifyModel;

use crate::common::Money;
use crate::wire::{append_len_str, read_len_str};

/// Channel is the `oneof` from the case file. Rust has a native sum type, so it
/// maps straight onto an enum — and because a sum type is exactly what `oneof`
/// describes, the same derive handles it. Variant names become schema tags in
/// snake_case; no converter, no marker trait, and no way to build a notification
/// carrying two targets at once.
#[derive(SerifyModel)]
pub enum Channel {
    Silent,         // arity 0 — a unit variant
    Sms(String),    // arity 1 — a scalar payload
    Push(u64),      // arity 1 — a payload that exceeds 2^53
    Invoice(Money), // arity N — a struct payload
}

#[derive(SerifyModel)]
pub struct NotificationRecord {
    notification_id: u32,
    channel: Channel,
    urgent: bool,
}

impl NotificationRecord {
    pub fn marshal(&self) -> Result<Vec<u8>, String> {
        let mut buf = Vec::new();
        buf.extend_from_slice(&self.notification_id.to_le_bytes());

        // The tag ordinal is the variant's position in the case file's oneof,
        // which is this enum's declaration order. The schema tag *names* are the
        // derive's business, and never appear here.
        match &self.channel {
            Channel::Silent => buf.push(0), // a unit variant is nothing but its tag
            Channel::Sms(s) => {
                buf.push(1);
                append_len_str(&mut buf, s);
            }
            Channel::Push(n) => {
                buf.push(2);
                buf.extend_from_slice(&n.to_le_bytes());
            }
            Channel::Invoice(m) => {
                buf.push(3);
                append_len_str(&mut buf, &m.currency);
                buf.extend_from_slice(&m.amount_minor.to_le_bytes());
            }
        }

        buf.push(self.urgent as u8);
        Ok(buf)
    }

    pub fn unmarshal(data: &[u8]) -> Result<Self, String> {
        if data.len() < 5 {
            return Err("truncated".into());
        }
        let notification_id = u32::from_le_bytes(data[0..4].try_into().unwrap());
        let mut p = 5usize;

        let channel = match data[4] {
            0 => Channel::Silent,
            1 => Channel::Sms(read_len_str(data, &mut p)?),
            2 => {
                if p + 8 > data.len() {
                    return Err("truncated".into());
                }
                let n = u64::from_le_bytes(data[p..p + 8].try_into().unwrap());
                p += 8;
                Channel::Push(n)
            }
            3 => {
                let currency = read_len_str(data, &mut p)?;
                if p + 8 > data.len() {
                    return Err("truncated".into());
                }
                let amount_minor = i64::from_le_bytes(data[p..p + 8].try_into().unwrap());
                p += 8;
                Channel::Invoice(Money {
                    currency,
                    amount_minor,
                })
            }
            ord => return Err(format!("unknown channel ordinal {ord}")),
        };

        if p >= data.len() {
            return Err("truncated".into());
        }
        Ok(Self {
            notification_id,
            channel,
            urgent: data[p] != 0,
        })
    }
}

File: examples/go/notification.go
// Copyright 2026 Chengxi Luo
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"encoding/binary"
	"fmt"
)

// Channel is the `sum` from examples/cases/notification.yaml. Go has no sum
// type, so the idiomatic encoding is a sealed interface: the unexported marker
// method means only the four types below can be a Channel — with a bare `any`,
// `n.Channel = 42` would compile.
//
// Sealing is all the marker does; the schema's tag names live in exactly one
// place, the Channel converter in main.go, and nothing in this file mentions them.
//
// What sealing does *not* buy is exhaustiveness: add a fifth variant and Go will
// not point at the switches below or at the converter. Rust's `match` would.
// That gap is real, and this example does not paper over it.
type Channel interface{ isChannel() }

type (
	Silent  struct{} // arity 0 — a unit variant
	SMS     string   // arity 1 — a scalar payload
	Push    uint64   // arity 1 — a payload that exceeds 2^53
	Invoice Money    // arity N — a struct payload
)

func (Silent) isChannel()  {}
func (SMS) isChannel()     {}
func (Push) isChannel()    {}
func (Invoice) isChannel() {}

// NotificationRecord mirrors examples/cases/notification.yaml.
type NotificationRecord struct {
	NotificationID uint32  `serify:"notification_id"`
	Channel        Channel `serify:"channel"`
	Urgent         bool    `serify:"urgent"`
}

func (n *NotificationRecord) MarshalBinary() ([]byte, error) {
	if n.Channel == nil {
		return nil, fmt.Errorf("notification %d has no channel", n.NotificationID)
	}
	buf := binary.LittleEndian.AppendUint32(nil, n.NotificationID)

	// The tag ordinal is the variant's position in the case file's sum, which
	// is the declaration order of the four types above.
	switch ch := n.Channel.(type) {
	case Silent:
		buf = append(buf, 0) // a unit variant is nothing but its tag
	case SMS:
		buf = append(buf, 1)
		buf = appendLenStr(buf, string(ch))
	case Push:
		buf = append(buf, 2)
		buf = binary.LittleEndian.AppendUint64(buf, uint64(ch))
	case Invoice:
		buf = append(buf, 3)
		buf = appendLenStr(buf, ch.Currency)
		buf = binary.LittleEndian.AppendUint64(buf, uint64(ch.AmountMinor))
	default:
		return nil, fmt.Errorf("notification %d: unhandled channel %T", n.NotificationID, ch)
	}

	if n.Urgent {
		buf = append(buf, 1)
	} else {
		buf = append(buf, 0)
	}
	return buf, nil
}

func (n *NotificationRecord) UnmarshalBinary(b []byte) error {
	if len(b) < 5 {
		return errTruncated
	}
	n.NotificationID = binary.LittleEndian.Uint32(b)
	ord := b[4]
	b = b[5:]

	switch ord {
	case 0:
		n.Channel = Silent{}
	case 1:
		s, rest, err := readLenStr(b)
		if err != nil {
			return err
		}
		n.Channel, b = SMS(s), rest
	case 2:
		if len(b) < 8 {
			return errTruncated
		}
		n.Channel, b = Push(binary.LittleEndian.Uint64(b)), b[8:]
	case 3:
		currency, rest, err := readLenStr(b)
		if err != nil {
			return err
		}
		if len(rest) < 8 {
			return errTruncated
		}
		n.Channel = Invoice{Currency: currency, AmountMinor: int64(binary.LittleEndian.Uint64(rest))}
		b = rest[8:]
	default:
		return fmt.Errorf("unknown channel ordinal %d", ord)
	}

	if len(b) < 1 {
		return errTruncated
	}
	n.Urgent = b[0] != 0
	return nil
}
